### PR TITLE
feat: catch imports of private definitions in resolution

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

Currently Noir only defines import visibility for functions, while everything else is implicitly public. This PR is an experiment on adding proper import visibility to Noir so that we can.

I've modified `ItemScope` so that it can hold both public and private definitions and during as we traverse modules in resolution we check that `dep::` paths don't enter any private modules. I've added `pub mod` syntax to allow this to be shown.

I've set the prelude to be `pub mod` so that we can compile most programs but all other modules within the stdlib are private to show how errors look like.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [x] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.
I'll document this closer to the time of this being merged.
# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: noir_stdlib/src/lib.nr
Comment (TomAFrench) [inline noir_stdlib/src/lib.nr:1]: Note that we're breaking a bunch of Rust visibility rules here. The prelude is importing things from private modules.

File: compiler/noirc_frontend/src/hir/resolution/import.rs
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/resolution/import.rs:321]: This is working around private modules not making their contents available to their parent module (i.e. importing public functions from a private module will fail).

Seems like we should add `pub(super)` functionality and make this the default for `mod foo;`

**Comment by TomAFrench** [comment]
Heads up that I'm going to be chopping this PR up into chunks to get some less potentially controversial sections through quicker and reduce the diff.

File: compiler/noirc_frontend/src/hir/resolution/import.rs
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/resolution/import.rs:215]: Us just erroring out here and not returning the fully resolved path _as well_ as the error makes this probably the most breaking change that we've ever done.

We're probably going to have to have this be a warning for a number of releases to avoid bricking most code out there.

File: compiler/noirc_frontend/src/parser/parser/function.rs
Comment (jfecher) [reply compiler/noirc_frontend/src/parser/parser/function.rs:61]: If we're already making breaking changes, I'd be fine with swapping it

Comment (TomAFrench) [reply compiler/noirc_frontend/src/parser/parser/function.rs:61]: I think we can hold off on that for now.

We can make both orders valid for a while alongside a formatter rule which prefers visibility first and then enforce that ordering at a later date.

**Comment by Savio-Sou** [comment]
Is this an extension of the [`open`](https://github.com/noir-lang/noir/pull/1037) and [`internal`](https://github.com/noir-lang/noir/pull/1873) keywords to general Noir programs?

**Comment by TomAFrench** [comment]
Nope this is an equivalent of https://doc.rust-lang.org/reference/visibility-and-privacy.html

File: compiler/noirc_frontend/src/hir/resolution/import.rs
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/resolution/import.rs:28]: Fix this

File: compiler/noirc_frontend/src/hir/resolution/import.rs
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/resolution/import.rs:19]:
```suggestion
pub enum PathResolution {
    Resolved {
      module_id: ModuleId,
      namespace: PerNs;
      errors: Vec<PathResolutionError>
    }
    Unresolved {
          errors: Vec<PathResolutionError>
    }
}
```

Thinking that we may want to refactor this into an enum to make it easier for us to pass back warnings/errors.

**Comment by TomAFrench** [comment]
I think that this is close enough to be ready for reviews now.

File: compiler/noirc_frontend/src/parser/parser.rs
Comment (TomAFrench) [inline compiler/noirc_frontend/src/parser/parser.rs:392]: This is a bit confusing as we're overloading `pub` for both zk visibiilty and standard programming language visibility.

An alternative would be to rename this to `export` so we'd have `export(super)` and `export(crate)` to avoid this overloading.

Comment (TomAFrench) [reply compiler/noirc_frontend/src/parser/parser.rs:392]: It's a break from rust but a relatively weak one as it's just a keyword change.

Comment (Savio-Sou) [reply compiler/noirc_frontend/src/parser/parser.rs:392]: > a break from rust
`pub` for ZK is a break anyway; the benefits of more intuitive keywords almost always outweigh the costs.

File: compiler/noirc_frontend/src/ast/mod.rs
Comment (jfecher) [inline compiler/noirc_frontend/src/ast/mod.rs:369]: Shouldn't `priv` print nothing? That way the formatter can use this function as well

File: compiler/noirc_frontend/src/ast/statement.rs
Comment (jfecher) [inline compiler/noirc_frontend/src/ast/statement.rs:253]: Ah, I see the check now. This should have a space after it though

File: compiler/noirc_frontend/src/hir/resolution/path_resolver.rs
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/resolution/path_resolver.rs:93]: Looks like mostly duplicated code from `import.rs`, can you refactor it to a common helper function?

File: compiler/noirc_frontend/src/lexer/token.rs
Comment (jfecher) [inline compiler/noirc_frontend/src/lexer/token.rs:760]:
```suggestion
```

File: compiler/noirc_frontend/src/parser/parser.rs
Comment (jfecher) [inline compiler/noirc_frontend/src/parser/parser.rs:373]:
```suggestion
/// import_visibility: 'pub(crate)' | 'pub' | %empty
```
The grammar with `?` would accept `pub(crate) pub`

File: compiler/noirc_frontend/src/parser/parser.rs
Comment (jfecher) [inline compiler/noirc_frontend/src/parser/parser.rs:405]:
```suggestion
        |(mut visibility, ident)| {
            // A module's contents are visible to the parent module unless they themselves are marked private.
            if visibility == ItemVisibility::Private {
                visibility = ItemVisibility::PublicSuper;
            }
            TopLevelStatement::Module(ModuleDeclaration { ident, visibility })
```

File: compiler/noirc_frontend/src/ast/statement.rs
Comment (TomAFrench) [reply compiler/noirc_frontend/src/ast/statement.rs:253]:
```suggestion
        if self.visibility != ItemVisibility::Private {
            write!(f, "{} ", self.visibility)?;
        };
```

**Review by jfecher — APPROVED** [review]
LGTM

File: compiler/noirc_frontend/src/parser/parser.rs
Comment (TomAFrench) [reply compiler/noirc_frontend/src/parser/parser.rs:392]: @jfecher what are your thoughts on this?

Comment (jfecher) [reply compiler/noirc_frontend/src/parser/parser.rs:392]: It sounds like using `pub` is a compromise either way. `pub` for item visibility is much more common within Noir while `pub` for zk is only used on entry-point function parameters. Changing either would be breaking but changing `pub` visibility may be less so just because it is newer. Using `export fn`, `export(crate) fn`, etc would make Noir less rust-like which could be a stumbling block. I'm fine with both uses of `pub` for now but if there's ever a future scenario where we could want to apply `pub (zk)` to a function somehow then things would get confusing real fast.

**Comment by Savio-Sou** [comment]
@TomAFrench would you like to supplement docs before merging the PR in?

**Comment by Savio-Sou** [comment]
Also will this be a breaking change in terms of libraries without `pub` will turn from public-by-default to private-by-default after the PR?

**Comment by TomAFrench** [comment]
Closing as superceded by more recent work.

**Pull request closed without merging**
